File: merge-git-projects.php
#!/usr/bin/env php
<?php
/***************************************************************
*  Copyright notice
*
*  (c) 2017 Dmitry Dulepov <dmitry.dulepov@gmail.com>
*  All rights reserved
*
*  This script is free software; you can redistribute it and/or modify
*  it under the terms of the GNU General Public License as published by
*  the Free Software Foundation; either version 3 of the License, or
*  (at your option) any later version.
*
*  The GNU General Public License can be found at
*  http://www.gnu.org/copyleft/gpl.html.
*
*  This script is distributed in the hope that it will be useful,
*  but WITHOUT ANY WARRANTY; without even the implied warranty of
*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
*  GNU General Public License for more details.
*
*  This copyright notice MUST APPEAR in all copies of the script!
***************************************************************/

/**
 * This class contains a tool to merge several git repositories to
 * a single repository preserving history.
 *
 * Credits:
 * - http://stackoverflow.com/a/14470212/2071628
 * - http://stackoverflow.com/a/39027521/2071628
 * - http://stackoverflow.com/a/4991675/2071628
 *
 * Requirements:
 * - PHP 5.6+
 * - git 2.0+
 * - Linux/Unix/OS X (not Windows!)
 * - bash or zsh as the main shell
 *
 * @author Dmitry Dulepov <dmitry.dulepov@gmail.com>
 */
class Merge {

    /** Set this to branches to be ignored in merged projects */
    const IGNORE_BRANCHES = '/origin\/(HEAD|master|dev)/';

    /**
     * Your desired git config for the merged project is here.
     *
     * @var array
     */
    protected $gitConfig = array();

    /**
     * Configuration of the main project.
     *
     * @var array
     */
    protected $mainProject = array();

    /**
     * Confugration for projects to merge.
     *
     * @var array
     */
    protected $projectsToMerge = array();

    /**
     * Tracker for created git branches.
     *
     * @internal
     * @var array
     */
    protected $createdLocalBranches = array();

    /**
     * Creates the instance of the class.
     */
    public function __construct() {
        if ($GLOBALS['argc'] < 2) {
            printf("Format: %s configfile.json\n", basename(__FILE__));
            exit(1);
        }

        $this->parseConfiguration();

        // Prevent merge commit message prompt
        putenv('GIT_MERGE_AUTOEDIT=no');
    }

    /**
     * Executes the merge.
     */
    public function execute() {
        $this->cloneRepository($this->mainProject['repository'], $this->mainProject['name'], $this->mainProject['mainBranch']);
        $this->configureRepository();
        $this->createNewMainBranch();
        foreach ($this->projectsToMerge as $projectName => $projectToMerge) {
            $this->cloneRepository($projectToMerge['repository'], $projectName, $projectToMerge['mainBranch']);
            $this->rewriteHistory($projectName);
            $this->findNonMergedBranches($projectName);
            $this->mergeProject($projectName);
        }
    }

    /**
     * Clones the repository.
     *
     * @param string $repository
     * @param string $subDirectory
     * @param string $mainBranch
     */
    protected function cloneRepository($repository, $subDirectory, $mainBranch) {
        if (file_exists(getcwd() . '/' . $subDirectory)) {
            $this->executeShellCommand('rm -fR %s', array(getcwd() . '/' . $subDirectory));
            echo "Removed directory: $subDirectory\n";
        }
        $this->executeShellCommand('git clone %s %s -b %s', array($repository, $subDirectory, $mainBranch));
    }

    /**
     * Configures the repository.
     */
    protected function configureRepository() {
        $currentDirectory = getcwd();
        chdir($currentDirectory . '/' . $this->mainProject['name']);
        foreach ($this->gitConfig as $key => $value) {
            $this->executeShellCommand('git config %s %s', array(escapeshellarg($key), escapeshellarg($value)));
        }
        chdir($currentDirectory);
    }

    /**
     * Creates a new main branch.
     */
    protected function createNewMainBranch() {
        $currentDirectory = getcwd();
        chdir($currentDirectory . '/' . $this->mainProject['name']);
        $this->executeShellCommand('git checkout -b %s', array($this->mainProject['createBranch']));
        chdir($currentDirectory);
    }

    /**
     * Runs emercency shell and prompts the user whether to continue or not.
     */
    protected function emergencyShell() {
        echo "Something went wrong. Bringing the emergency shell to correct errors manually...\n";
        echo "===========================================\n\n";

        $shell = getenv('SHELL');
        exec($shell . ' -l');

        echo "\n\n===========================================\n\n";
        echo "Emergency shell finished. ";
        do {
            echo "Continue (y/n)? ";
            $answer = strtolower(trim(fgets(STDIN)));
        } while ($answer !== 'y' && $answer !== 'n');

        if ($answer === 'n') {
            exit(1);
        }
    }

    /**
     * Executes shell command and returns the last line.
     *
     * @param string $format
     * @param array $parameters
     * @param array $output
     * @return string
     */
    protected function executeShellCommand($format, array $parameters = array(), array &$output = null) {
        $exitCode = 0;
        $output = array();

        $command = vsprintf($format, $parameters);

        echo "Executing: $command\n";

        $command .=  ' 2>&1';
        exec(getenv('SHELL') . ' -c ' . escapeshellarg($command), $output, $exitCode);
        $lastLine = end($output);
        if ($exitCode != 0) {
            foreach ($output as $line) {
                print("$line\n");
            }
            $this->emergencyShell();
        }

        return trim($lastLine);
    }

    /**
     * Finds non-merged branches in the projects, checks them out and saves for later use.
     *
     * WARNING: this must be called after history rewrite!
     *
     * @param string $projectName
     */
    protected function findNonMergedBranches($projectName) {
        $output = array();
        $project = &$this->projectsToMerge[$projectName];
        $currentDirectory = getcwd();
        chdir($currentDirectory . '/' . $projectName);
        $this->executeShellCommand('git branch -r --no-merged %s', array($project['mainBranch']), $output);

        $branchStartingPointCommand = "diff -u <(git rev-list --first-parent %s) <(git rev-list --first-parent %s) | sed -ne 's/^ //p' | head -1";
        foreach ($output as $remoteBranch) {
            $remoteBranch = trim($remoteBranch);
            $regexp = '/' . str_replace('/', '\/', $project['ignoreBranches']) . '/';
            if (($project['ignoreBranches'] === '' || !preg_match($regexp, $remoteBranch)) && substr($remoteBranch, 0, 7) === 'origin/') {
                $localBranch = substr($remoteBranch, 7);
                $this->executeShellCommand('git checkout -b %s %s', array($localBranch, $remoteBranch));
                $project['copyBranches'][$localBranch] = $this->executeShellCommand($branchStartingPointCommand, array($localBranch, $project['mainBranch']));
            }
        }
        $this->executeShellCommand('git checkout %s', array($project['mainBranch']));
        chdir($currentDirectory);
    }

    /**
     * Merges the specified project into the new project.
     *
     * @param string $projectName
     */
    protected function mergeProject($projectName) {
        $project = $this->projectsToMerge[$projectName];

        $currentDirectory = getcwd();
        chdir($currentDirectory . '/' . $this->mainProject['name']);

        $this->executeShellCommand('git remote add -f %1$s ../%1$s', array($projectName));
        $this->executeShellCommand('git merge --no-ff %s/%s --allow-unrelated-histories', array($projectName, $project['mainBranch']));
        foreach ($project['copyBranches'] as $branchName => $commitId) {
            if (!isset($this->createdLocalBranches[$branchName])) {
                $this->executeShellCommand('git checkout -b %s %s', array($branchName, $commitId));
                $this->createdLocalBranches[$branchName] = 1;
            }
            else {
                echo "Warning: merging into existing local branch ($branchName)!\n";
                $this->executeShellCommand('git checkout %s', array($branchName));
            }
            $this->executeShellCommand('git merge --no-ff %s/%s --allow-unrelated-histories', array($projectName, $branchName));
        }
        // Reset to the newly created branch
        $this->executeShellCommand('git checkout %s', array($this->mainProject['createBranch']));

        $this->executeShellCommand('git remote remove %s', array($projectName));

        chdir($currentDirectory);

        $this->executeShellCommand('rm -fR %s', array($projectName));
    }

    /**
     * Parses and vaidates the configuration.
     */
    protected function parseConfiguration() {
        $configrationFile = $GLOBALS['argv'][1];
        $jsonString = @file_get_contents($configrationFile);
        if (!$jsonString) {
            printf("File %s cannot be read\n", $configrationFile);
            exit(1);
        }
        $configuration = json_decode($jsonString, true);
        if (!is_array($configuration)) {
            printf("File %s does not contain valid json\n", $configrationFile);
            exit(1);
        }

        $this->mainProject = (array)$configuration['mainProject'];
        $this->projectsToMerge = (array)$configuration['projectsToMerge'];
        $this->gitConfig = (array)$configuration['gitConfig'];

        // Validate
        foreach (array('name', 'repository', 'mainBranch', 'createBranch') as $key) {
            if (!isset($this->mainProject[$key])) {
                printf("Mandatory key '$key' is missing in the 'mainProject' configuration\n");
                exit(1);
            }
        }
        foreach ($this->projectsToMerge as $name => &$projectConfiguration) {
            foreach (array('repository', 'path', 'mainBranch', 'ignoreBranches') as $key) {
                if (!isset($projectConfiguration[$key])) {
                    printf("Mandatory key '$key' is missing in the '$name' project configuration\n");
                    exit(1);
                }
            }
            $projectConfiguration['copyBranches'] = array();
        }
    }

    /**
     * Rewrites history to place the files from the subproject to a subdirectory
     * like in the main project.
     *
     * @param $projectName
     */
    protected function rewriteHistory($projectName) {
        $project = $this->projectsToMerge[$projectName];
        $currentDirectory = getcwd();
        chdir($currentDirectory . '/' . $projectName);

        $command = "git filter-branch -f --prune-empty --tree-filter \"mkdir -p %s ; git ls-tree --name-only \\\$GIT_COMMIT | xargs -I file mv file %s\" -- --all";
        $this->executeShellCommand($command, array($project['path'], $project['path']));

        chdir($currentDirectory);
    }
}

if (PHP_SAPI === 'cli') {
    (new Merge())->execute();
}

File: merge-git-projects.py
import argparse
import json
import os
import pty
import re
import shlex
import shutil
import subprocess
import sys

"""
This file contains a tool to merge several git repositories to
a single repository preserving history.

Credits:
- http://stackoverflow.com/a/14470212/2071628
- http://stackoverflow.com/a/39027521/2071628
- http://stackoverflow.com/a/4991675/2071628

Requirements:
- Python 3.5+
- git 2.0+
- Linux/Unix/OS X (not Windows!)
- bash or zsh as the main shell
- zsh

Author: Dmitry Dulepov <dmitry.dulepov@gmail.com>
"""
class MergeGitProjects:

    def __init__(self):
        """
        Initializes the class.
        """
        self.__configuration = {}
        self.__created_local_branches = []
        self.__verbose = False
        # Disable merge message editing
        os.putenv('GIT_MERGE_AUTOEDIT', 'no')


    def execute(self):
        """
        Executes the merge
        """
        self._parse_arguments()

        print("Creating a copy of the main repository...")

        self._clone_repository(self.__configuration['mainProject']['repository'],
                self.__configuration['mainProject']['name'], self.__configuration['mainProject']['mainBranch'])
        self._configure_repository()
        self._create_new_main_branch()

        for project_name in self.__configuration['projectsToMerge']:
            print("Merging project '%s'..." % project_name)
            project_to_merge = self.__configuration['projectsToMerge'][project_name]
            self._clone_repository(project_to_merge['repository'], project_name, project_to_merge['mainBranch'])
            self._rewrite_history(project_name)
            self._find_non_merged_branches(project_name)
            self._merge_project(project_name)


    def _clone_repository(self, repository, directory_name, main_branch):
        """
        Clones the remote repository in the subdirectory and checks out the given branch
        
        :param repository: repository specification 
        :param directory_name: subdirectory name (relative to current) where to clone
        :param main_branch: branch to checkout
        """

        def directory_was_not_removed():
            print('Error: could not remove %s' % directory_name)
            sys.exit(1)

        current_directory = os.getcwd()
        repository_directory = os.path.join(current_directory, directory_name)

        if os.path.exists(repository_directory):
            if self.__verbose:
                print('Removing directory %s' % directory_name)
            shutil.rmtree(directory_name, onerror=directory_was_not_removed)

        self._execute_shell_command('git clone %s %s -b %s', repository, directory_name, main_branch)


    def _configure_repository(self):
        """
        Configures the main repository according to the configuration.
        """
        current_directory = os.getcwd()
        os.chdir(os.path.join(current_directory, self.__configuration['mainProject']['name']))

        for key in self.__configuration['gitConfig']:
            self._execute_shell_command('git config %s %s', shlex.quote(key), shlex.quote(self.__configuration['gitConfig'][key]))

        os.chdir(current_directory)


    def _create_new_main_branch(self):
        """
        Creates a new branch in the main repository
        """
        current_directory = os.getcwd()
        os.chdir(os.path.join(current_directory, self.__configuration['mainProject']['name']))
        self._execute_shell_command('git checkout -b %s', self.__configuration['mainProject']['createBranch'])
        os.chdir(current_directory)


    def _emergency_shell(self):
        """
        Runs emergency shell
        """
        print("Something went wrong. Bringing the emergency shell to correct errors manually...\n")
        print("===========================================\n\n")

        pty.spawn([os.getenv('SHELL'), '-l'])

        print("\n\n===========================================\n\n")
        print("Emergency shell finished. ")
        answer = ''
        while answer != 'y' and answer != 'n':
            answer = input("Continue (y/n)? ")

        if answer == 'n':
            sys.exit(1)


    def _execute_shell_command(self, command_format, *args):
        """
        Runs the shell with the given command

        :param command_format: command to run (like 'git checkout %s') 
        :param args: arguments to format the command
        :return: array of lines from stdout
        """
        command = str(command_format) % args

        if self.__verbose:
            print('Executing: %s' % command)

        result = subprocess.run(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=True, executable=os.getenv('SHELL'), universal_newlines=True)
        if result.returncode != 0:
            self._emergency_shell()

        return result.stdout.splitlines()


    def _find_non_merged_branches(self, project_name):
        """
        Finds branches in the project that are not merged to the project's main branch.
        :param: project_name: project name to search 
        """
        new_branches = {}
        project = self.__configuration['projectsToMerge'][project_name]
        branch_starting_point_command = "diff -u <(git rev-list --first-parent %s) <(git rev-list --first-parent dev) | sed -ne 's/^ //p' | head -1"

        current_directory = os.getcwd()
        os.chdir(os.path.join(current_directory, project_name))
        output = self._execute_shell_command('git branch -r --no-merged %s', project['mainBranch'])

        for remote_branch in output:
            remote_branch = remote_branch.strip()
            if project['ignoreBranches'] == '' or not re.match(project['ignoreBranches'], remote_branch) and remote_branch[0:7] == 'origin/':
                local_branch = remote_branch[7:]
                self._execute_shell_command('git checkout -b %s %s', local_branch, remote_branch)
                new_branches[local_branch] = self._execute_shell_command(branch_starting_point_command, local_branch)[0]

        self._execute_shell_command('git checkout %s', project['mainBranch'])

        os.chdir(current_directory)

        self.__configuration['projectsToMerge'][project_name]['copyBranches'] = new_branches


    def _merge_project(self, project_name):
        """
        Merges the project

        :param project_name: project name from the configuration 
        """
        project = self.__configuration['projectsToMerge'][project_name]

        current_directory = os.getcwd()
        os.chdir(os.path.join(current_directory, self.__configuration['mainProject']['name']))

        self._execute_shell_command('git remote add -f %s ../%s', project_name, project_name)
        self._execute_shell_command('git merge --no-ff %s/%s --allow-unrelated-histories', project_name, project['mainBranch'])
        for branch_name in project['copyBranches']:
            commit_id = project['copyBranches'][branch_name]
            if branch_name not in self.__created_local_branches:
                self._execute_shell_command('git checkout -b %s %s', branch_name, commit_id)
                self.__created_local_branches.append(branch_name)
            else:
                print("Warning: merging into existing local branch (%s)!" % branch_name)
                self._execute_shell_command('git checkout %s', branch_name)
            self._execute_shell_command('git merge --no-ff %s/%s --allow-unrelated-histories', project_name, branch_name)

        # Reset to the newly created branch
        self._execute_shell_command('git checkout %s', self.__configuration['mainProject']['createBranch'])

        self._execute_shell_command('git remote remove %s', project_name)

        os.chdir(current_directory)

        shutil.rmtree(os.path.join(current_directory, project_name))


    def _parse_arguments(self):
        """
        Parses arguments
        """
        parser = argparse.ArgumentParser(
            add_help=False,
            description='This script merges one or more git projects into a base projects preserving history.',
            epilog='See https://github.com/dmitryd/merge-git-projects for more information.'
        )
        parser.add_argument('configuration_file_name', metavar='configuration-file-name', help='Configuration file', default='')
        parser.add_argument('-v', action='store_true', dest='verbose', help='Show what is done')
        arguments = parser.parse_args()

        self.__verbose = arguments.verbose

        try:
            configuration_file = open(arguments.configuration_file_name, 'r')
            self.__configuration = json.load(configuration_file)
            configuration_file.close()
        except FileNotFoundError:
            print('Error: configuration file "%s" not found' % arguments.configuration_file_name)
            sys.exit(1)

        for section in ['gitConfig', 'mainProject', 'projectsToMerge']:
            if section not in self.__configuration:
                print('Error: "%s" section is missing in the configuration file' % section)
                sys.exit(1)

        for option in ['name', 'repository', 'mainBranch', 'createBranch']:
            if option not in self.__configuration['mainProject']:
                print('Error: "%s" option is missing in the "mainProject" section in the configuration file' % option)
                sys.exit(1)

        for projectName in self.__configuration['projectsToMerge']:
            project = self.__configuration['projectsToMerge'][projectName]
            for option in ['repository', 'path', 'mainBranch', 'ignoreBranches']:
                if option not in project:
                    print('Error: "%s" option is missing in the "%s" project in the configuration file' %
                          (option, projectName))
                    sys.exit(1)


    def _rewrite_history(self, project_name):
        """
        Retwrites history for the project. This moves the project into its right
        place inside the main project.

        :param: project_name: project name to process
        """
        project = self.__configuration['projectsToMerge'][project_name]
        current_directory = os.getcwd()
        os.chdir(os.path.join(current_directory, project_name))

        command = "git filter-branch -f --tree-filter \"zsh -c 'setopt extended_glob && setopt glob_dots && mkdir -p %s && (mv ^(.git|%s) %s || true)'\" -- --all"
        first_dir = project['path'].split(os.path.sep)[0:1][0]
        self._execute_shell_command(command, project['path'], first_dir, project['path'])

        os.chdir(current_directory)


if __name__ == "__main__":
    merge = MergeGitProjects()
    merge.execute()
